src/cephadm/cephadm

```diff
@@ -3450,26 +3453,144 @@ def command_list_networks():
+    try:
+        units = os.listdir(f'/etc/systemd/system/ceph-{fsid}.target.wants')
+    except FileNotFoundError:
+        return []
+    else:
+        return units
+
 
-    data_dir = args.data_dir
+def get_unit_states(fsid: str) -> Dict[str, Dict[str, str]]:
+    cmd = f'systemctl --all --no-legend -t service list-units ceph-{fsid}*'
+    unit_states = {}
+    std_out, std_err, rc = call(cmd.split(' '), verbose_on_failure=False)
+    if not rc:
+        for line in std_out.split('\n'):
+            if not line:
+                continue
+            unit_name, load, active, sub = line.split()[0:4]
+            unit_states[unit_name] = {
+                "load": load,
+                "active": active,
+                "sub": sub,
+            }
+    return unit_states
+
+
+def list_daemons(detail=True, legacy_dir=None, target_fsid=None):
+    # type: (bool, Optional[str], Optional[str]) -> List[Dict[str, str]]
+    host_version: Optional[str] = None
+    ls: List[Dict[str, Optional[Union[str, bool]]]] = []
+    fsid: str
+    info: Dict[str, Optional[Union[str, bool]]]
+
+
+    def _http_call(url: str, context: Optional[ssl.SSLContext] = None) -> str:
+        try:
+            with urlopen(url, context=context) as response:
+                return response.read().decode('utf-8')
+        except URLError:
+            return ''
+
+    def _get_listen_address(file_name: str) -> Optional[str]:
+        with open(file_name) as f:
+            data = f.read().strip()
+        if "web.listen-address" in data:
+            listen_address = data.split('--web.listen-address=')[1].split(' ')[0]
+            return listen_address.split(':')[-1]
+        return None
+
+    def _get_grafana_port(file_name: str) -> Optional[str]:
+        with open(file_name) as f:
+            data = f.read().strip()
+        for line in data.split('\n'):
+            if 'http_port' in line and line.startswith(' '):
+                return line.split('=')[1].strip()
+        return None
+
+    def _get_web_port(service: str) -> Optional[str]:
+        dirs = os.listdir(os.path.join(data_dir, fsid))
+        for dir_name in dirs:
+            if dir_name.startswith(service):
+                if service in ['node-exporter', 'alertmanager', 'prometheus']:
+                    return _get_listen_address(os.path.join(data_dir, fsid, dir_name, 'unit.run'))
+                elif service == 'grafana':
+                    return _get_grafana_port(os.path.join(data_dir, fsid, dir_name, 'etc', 'grafana', 'grafana.ini'))
+                break
+        return None
+
+    def _get_grafana_version(default_port: str = '3000') -> str:
+        port = _get_web_port('grafana') or default_port
+        ctx = ssl.create_default_context()
+        ctx.check_hostname = False
+        ctx.verify_mode = ssl.CERT_NONE
+        data = _http_call(
+            f"https://localhost:{port}/api/health",
+            ctx)
+        return json.loads(data)['version']
```

Comment:
Yes - and the process is the time consuming thing. Take prometheus as an example
[root@rhs-srv-01 ~]# time curl http://localhost:9095/api/v1/status/buildinfo
{"status":"success","data":{"version":"2.18.1","revision":"ecee9c8abfd118f139014cb1b174b08db3f342cf","branch":"HEAD","buildUser":"root@2117a9e64a7e","buildDate":"20200507-16:51:47","goVersion":"go1.14.2"}}
real    0m0.015s
user    0m0.010s
sys     0m0.005s
[root@rhs-srv-01 ~]# time podman exec -it 6d92753bd8a8 prometheus --version
prometheus, version 2.18.1 (branch: HEAD, revision: ecee9c8abfd118f139014cb1b174b08db3f342cf)
  build user:       root@2117a9e64a7e
  build date:       20200507-16:51:47
  go version:       go1.14.2

real    0m0.907s
user    0m0.139s
sys     0m0.135s

That's nearly a factor of 10 slower using exec - this is why I'm trying to avoid it where possible, and add the ceph version centrally on the manager.

Reply:
I for I got was:

> `podman exec` involves (without going into to much detail) spinning up `podman` (including reading a couple of config files), spinning up `conmon` and `runc` before executing the final command. I can't really say how much that overhead is, but I can image that it is significant for running a short command like that.

I think we can happily continue with `curl` here.

Reply:
done